src/result.rs

```diff
@@ -42,22 +42,26 @@ pub enum Error {
     #[doc(hidden)]
     #[cfg(feature = "failpoints")]
     FailPoint,
+    /// Compare and swap failure.
+    CompareAndSwap {
+        /// Current value if present.
+        cur: Option<IVec>,
+    },
```

Comment:
CompareAndSwap should not be a member of this Error enum, because it represents a high-level normal failure and not an operational critical error. In the past, these were mixed, and it led to a lot of bugs because it basically makes exhaustive pattern matching useless on Error types. It's better to use nested result types when dealing with multiple levels of concerns, even though it feels more heavy weight. It pushes the real complexity at play into awareness of the user.